claat/parser/md/parse.go

```diff
@@ -0,0 +1,507 @@
+	metaStatus           = "status"
+	metaFeedbackLink     = "feedback link"
+	metaAnalyticsAccount = "analytics account"
+)
+
+var metadataRegexp = regexp.MustCompile(`(.+?):(.+)`)
+var languageRegexp = regexp.MustCompile(`language-(.+)`)
+var durationHintRegexp = regexp.MustCompile(`(?i)Duration:? (.+)`)
+var durationRegexp = regexp.MustCompile(`(\d+)[:.](\d{2})$`)
+var downloadButtonRegexp = regexp.MustCompile(`^(?i)Download(.+)$`)
+
+// init registers this parser so it is available to CLaaT.
+func init() {
+	parser.Register("md", Parse)
+}
+
+// parserState encapsulates the state of the parser at any given step.
+type parserState struct {
+	tzr *html.Tokenizer
+	c   *types.Codelab
+	t   html.Token
+
+	currentStep *types.Step
+}
+
+// emit accepts a node, and either writes the node directly to the current step, or writes the node to the node buffer.
+func (ps *parserState) emit(n types.Node) {
+	ps.currentStep.Content.Append(n)
+}
+
+// advance moves the tokenizer to the next token and updates the token convenience variable.
+func (ps *parserState) advance() {
+	ps.tzr.Next()
+	ps.t = ps.tzr.Token()
+}
+
+// multiAdvance is a convenience method for repeatedly advancing.
+func (ps *parserState) multiAdvance(n int) {
+	for i := 0; i < n; i++ {
+		ps.advance()
+	}
+}
+
+// claatMarkdown calls the Blackfriday Markdown parser with some special addons selected. It takes a byte slice as a parameter,
+// and returns its result as a byte slice.
+func claatMarkdown(b []byte) []byte {
+	extns := blackfriday.EXTENSION_FENCED_CODE |
+		blackfriday.EXTENSION_NO_EMPTY_LINE_BEFORE_BLOCK |
+		blackfriday.EXTENSION_NO_INTRA_EMPHASIS |
+		blackfriday.EXTENSION_DEFINITION_LISTS |
+		blackfriday.EXTENSION_TABLES
+	o := blackfriday.Options{
+		Extensions: extns,
+	}
+	r := blackfriday.HtmlRenderer(extns, "", "")
+	return blackfriday.MarkdownOptions(b, r, o)
+}
+
+// Parse is a parser.ParseFunc for Markdown.
+func Parse(r io.Reader) (*types.Codelab, error) {
+	// Convert Markdown to HTML for easy parsing.
+	mBuf := bytes.Buffer{}
+	_, err := mBuf.ReadFrom(r)
+	if err != nil {
+		return nil, err
+	}
+	hBytes := claatMarkdown(mBuf.Bytes())
+	h := bytes.NewBuffer(hBytes)
+	// Parse the markup.
+	c, err := parseMarkup(h)
+	if err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// parseMarkup accepts an io.Reader to markup created by the Devsite Markdown parser. It returns a pointer to a codelab object, or an error if one occurs.
+func parseMarkup(markup io.Reader) (*types.Codelab, error) {
+	// Avoid global vars by encapsulating state.
+	ps := parserState{
+		tzr: html.NewTokenizer(markup),
+		c:   &types.Codelab{},
+	}
+
+	var inStepTitle bool
+
+	// Advance through the tokenized input, one token at a time.
+	for ps.advance(); ps.t.Type != html.ErrorToken; ps.advance() {
+		if ps.c.Title == "" {
+			// If we have a <p> tag, it's a metadata section.
+			if ps.t.Type == html.StartTagToken && ps.t.DataAtom == atom.P {
+				parseMetadata(&ps)
```

Comment:
You don't seem to be checking for errors here. Why bother returning the error from `parseMetadata` then?